cpp/src/arrow/compute/kernels/row_encoder_internal.h

```diff
@@ -424,49 +469,228 @@ struct ListFixedWidthChildEncoder : KeyEncoder {
+    for (int64_t i = 0; i < length; i++) {
+      auto encoded_ptr = encoded_bytes[i];
+      Offset list_length = util::SafeLoadAs<Offset>(encoded_bytes[i]);
+      encoded_ptr += sizeof(Offset);
+      for (int64_t j = 0; j < list_length; j++) {
+        ChildOffset child_length =
+          util::SafeLoadAs<ChildOffset>(encoded_ptr + kExtraByteForNull);
+        child_length_sum += child_length;
+        encoded_ptr += kExtraByteForNull + sizeof(ChildOffset) + child_length;
+      }
+      list_length_sum += list_length;
     }
-    raw_offsets[length] = current_child_offset;
-    writer.Finish();
+
+    ARROW_ASSIGN_OR_RAISE(auto offset_buf,
+                          AllocateBuffer(sizeof(Offset) * (1 + length), pool));
+    ARROW_ASSIGN_OR_RAISE(auto child_key_buf, AllocateBuffer(child_length_sum, pool));
+    ARROW_ASSIGN_OR_RAISE(auto child_validity_buf,
+                          AllocateBitmap(child_length_sum, pool));
+    ARROW_ASSIGN_OR_RAISE(auto child_offset_buf,
+                          AllocateBuffer(sizeof(ChildOffset) * (1 + list_length_sum)));
+
+    auto raw_offset = reinterpret_cast<Offset*>(offset_buf->mutable_data());
+    auto raw_child_key = child_key_buf->mutable_data();
+    auto raw_child_validity = child_validity_buf->mutable_data();
+    auto raw_child_offset =
+      reinterpret_cast<ChildOffset*>(child_offset_buf->mutable_data());
+
+    int64_t child_null_count = 0;
+    ChildOffset child_length_cumsum = 0;
+    ListDecoder<T>(encoded_bytes, length, child_length_sum,
+      child_null_count, raw_offset, raw_child_validity,
+    [&](int64_t i, int64_t){
+      ChildOffset child_length = util::SafeLoadAs<ChildOffset>(encoded_bytes[i]);
+      encoded_bytes[i] += sizeof(ChildOffset);
+
+      memcpy(raw_child_key, encoded_bytes[i], child_length);
+      raw_child_key += child_length;
+      encoded_bytes[i] += child_length;
+
+      util::SafeStore<ChildOffset>(raw_child_offset, child_length_cumsum);
+      child_length_cumsum += child_length;
+      raw_child_offset++;
+    }, [&](int64_t i, int64_t){
+      encoded_bytes[i] += sizeof(ChildOffset);
+      util::SafeStore<ChildOffset>(raw_child_offset, child_length_cumsum);
+      raw_child_offset++;
+    });
+    util::SafeStore<Offset>(raw_child_offset, child_length_cumsum);
 
     std::vector<std::shared_ptr<ArrayData>> child_data = {ArrayData::Make(
-      std::static_pointer_cast<T>(type_)->value_type(), length_sum,
-        {std::move(child_validity_buf), std::move(child_value_buf)}, child_null_count)};
+      std::static_pointer_cast<T>(type_)->value_type(), list_length_sum,
+        {std::move(child_validity_buf), std::move(child_offset_buf),
+          std::move(child_key_buf)}, child_null_count)};
 
     return ArrayData::Make(
       type_, length, {std::move(null_buf), std::move(offset_buf)},
-        child_data, null_count);
+      child_data, null_count);
   }
 
   std::shared_ptr<DataType> type_;
+  int64_t scalar_buffer_length;
```

Comment:
Yeah I agree that it is a poor design, but the tradeoff is that we have to go through the scalar list twice instead of once when encoding.

Though it should not be that big of a performance diff and I'll replace it with
`getListBufferSize(0, list_length, scalar_validity_buffer,
                                          scalar_offset_buffer)`